lib/inputs/dynamodb-data.js

```diff
@@ -0,0 +1,156 @@
+  encodeSegment(rec) {
+    const epochMs = (rec.timestamp || Date.now()).toString();
+    if (rec.type === 'segmentbytes') {
+      return `${epochMs}.${rec.segment}`;
+    } else {
+      return epochMs;
+    }
+  }
+
+  decodeSegment(str) {
+    const parts = str.split('.');
+    const epoch = safeParseInt(parts[0]);
+    const segment = safeParseInt(parts[1]) || OVERALL_DOWNLOAD;
+    return [timestamp.toDateString(epoch), epoch, segment];
+  }
+
+  async insert() {
+    const allKeys = Object.keys(this.payloads).concat(Object.keys(this.segments));
+    const uniqueKeys = allKeys.filter((v, i, a) => a.indexOf(v) === i);
+
+    if (uniqueKeys.length == 0) {
+      return Promise.resolve([]);
+    } else {
+      const updates = uniqueKeys.map(k => [k, this.payloads[k], this.segments[k]]);
+      const result = await dynamo.updateAll(updates);
+      const output = [{ dest: 'dynamodb', count: uniqueKeys.length }];
+
+      // send "new" segments onwards to kinesis
+      const formatted = result.success.map(r => this.format(r)).flat();
+      if (formatted.length) {
+        const count = await kinesis.put(formatted);
+        output.push({ count, dest: `kinesis:${kinesis.stream()}` });
+      }
+
+      // just throw an error if anything failed. successful records have already
+      // been recorded in DDB, so retrying them won't cause duplicates.
+      if (result.failures.length > 0) {
+        const fails = result.failures.length;
+        const total = result.failures.length + result.success.length;
+        const msg = `DDB retrying for ${fails}/${total}`;
+        logger.warn(msg, { ddb: 'retrying', count: fails });
+        throw new Error(msg);
+      }
+
+      return output;
+    }
+  }
+
+  // create combined "postbytes" records, when we have both both the initial
+  // redirect-data and the CDN segments-downloaded
+  format([key, payload, segments]) {
+    if (!payload || !segments) {
+      return [];
+    }
+
+    // for each UTC day, construct a record and filter download/impressions
+    const [todoSegments, todoTimestamps] = this.dedupSegments(segments);
+    return Object.keys(todoSegments).map(day => {
+      const record = this.decodeKey(key, payload);
+      record.type = record.type === 'antebytespreview' ? 'postbytespreview' : 'postbytes';
+      record.timestamp = todoTimestamps[day];
+
+      // check if this should be an overall download
+      if (!todoSegments[day].includes(OVERALL_DOWNLOAD)) {
+        delete record.download;
+      }
+
+      // filter segment impressions
+      if (record.impressions) {
+        record.impressions = record.impressions.filter(i =>
+          todoSegments[day].find(s => `${i.segment}` === `${s}`),
+        );
+      }
+
+      return record;
+    });
+  }
+
+  // we may have already processed a UTC-date + segment, and recieve newer
+  // timestamps for that same day. dedup them so you can't get multiple
+  // downloads/impressions on the same UTC day.
+  dedupSegments(segments) {
```

Comment:
This bit gets a bit confusing ... we need to dedup based on the UTC day.  But it's perfectly legal to download this `<listener_episode>.<digest>` twice, just before/after the UTC midnight boundary.  So we're making a list of segments to log here: `{ "2021-06-21: [1, 3, "DOWNLOAD"] }`.